PubSubToDatadog: Integration Tests

ash-ddog opened the pull request:
Follow-on to: https://github.com/GoogleCloudPlatform/DataflowTemplates/pull/790, will rebase off `main` once it goes in.

This PR adds integration tests for the `PubSubToDatadog` classic template. They work very similar to other integration tests in the repo. The one thing worth mentioning is that Datadog is SaaS-only, and so no containerized version of the software is available. So these tests instead spin up a [MockServer](https://www.mock-server.com/) container with request expectations that mimic the [Datadog Logs API](https://docs.datadoghq.com/api/latest/logs/#send-logs) and verify the recorded requests.

Let me know if you have any question!

ash-ddog commented on it/datadog/src/main/java/com/google/cloud/teleport/it/datadog/DatadogLogEntry.java:
```diff
@@ -0,0 +1,100 @@
+
+/** A class for Datadog log entries, copy of DatadogEvent. */
+@AutoValue
+public abstract class DatadogLogEntry {
```
This should be replaced by `DatadogEvent` when we get around to contributing the `DatadogIO` to Apache Beam.

bvolpato commented:
Thanks! This looks really good from a first glance, but it's a long PR and I will need some time to take a careful look.

bvolpato commented:
Thanks! Please check build here

```
/home/runner/actions-runner/_work/DataflowTemplates/DataflowTemplates/it/datadog/src/main/java/com/google/cloud/teleport/it/datadog/DatadogResourceManager.java:[285,7] constructor Builder in class com.google.cloud.teleport.it.testcontainers.TestContainerResourceManager.Builder<T> cannot be applied to given types;
  required: java.lang.String,java.lang.String,java.lang.String
  found: java.lang.String
  reason: actual and formal argument lists differ in length

```

bvolpato commented:
Got a 

```
2023-07-18T22:09:53.1687673Z java.lang.reflect.InvocationTargetException
2023-07-18T22:09:53.1688202Z 	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
2023-07-18T22:09:53.1688840Z 	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
2023-07-18T22:09:53.1689537Z 	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
2023-07-18T22:09:53.1689926Z 	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
2023-07-18T22:09:53.1690702Z 	at com.google.cloud.teleport.it.gcp.dataflow.DirectRunnerClient$DirectRunnerJobThread.run(DirectRunnerClient.java:214)
2023-07-18T22:09:53.1691209Z Caused by: java.lang.IllegalArgumentException: inputBatchCount must be greater than or equal to 10
2023-07-18T22:09:53.1691935Z 	at org.apache.beam.vendor.guava.v26_0_jre.com.google.common.base.Preconditions.checkArgument(Preconditions.java:216)
2023-07-18T22:09:53.1692650Z 	at com.google.cloud.teleport.datadog.DatadogEventWriter$Builder.withInputBatchCount(DatadogEventWriter.java:463)
2023-07-18T22:09:53.1693200Z 	at com.google.cloud.teleport.datadog.DatadogIO$Write.expand(DatadogIO.java:86)
2023-07-18T22:09:53.1693719Z 	at com.google.cloud.teleport.datadog.DatadogIO$Write.expand(DatadogIO.java:60)
2023-07-18T22:09:53.1694216Z 	at org.apache.beam.sdk.Pipeline.applyInternal(Pipeline.java:545)
2023-07-18T22:09:53.1695084Z 	at org.apache.beam.sdk.Pipeline.applyTransform(Pipeline.java:496)
2023-07-18T22:09:53.1695584Z 	at org.apache.beam.sdk.values.PCollection.apply(PCollection.java:365)
2023-07-18T22:09:53.1696152Z 	at com.google.cloud.teleport.templates.PubSubToDatadog.run(PubSubToDatadog.java:222)
2023-07-18T22:09:53.1696847Z 	at com.google.cloud.teleport.templates.PubSubToDatadog.main(PubSubToDatadog.java:150)
2023-07-18T22:09:53.1697010Z 	... 5 more

```

In the smoke tests, going to rebase/rerun to double check

ash-ddog commented:
> Got a
> 
> ```
> 2023-07-18T22:09:53.1687673Z java.lang.reflect.InvocationTargetException
> 2023-07-18T22:09:53.1688202Z 	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
> 2023-07-18T22:09:53.1688840Z 	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
> 2023-07-18T22:09:53.1689537Z 	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
> 2023-07-18T22:09:53.1689926Z 	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
> 2023-07-18T22:09:53.1690702Z 	at com.google.cloud.teleport.it.gcp.dataflow.DirectRunnerClient$DirectRunnerJobThread.run(DirectRunnerClient.java:214)
> 2023-07-18T22:09:53.1691209Z Caused by: java.lang.IllegalArgumentException: inputBatchCount must be greater than or equal to 10
> 2023-07-18T22:09:53.1691935Z 	at org.apache.beam.vendor.guava.v26_0_jre.com.google.common.base.Preconditions.checkArgument(Preconditions.java:216)
> 2023-07-18T22:09:53.1692650Z 	at com.google.cloud.teleport.datadog.DatadogEventWriter$Builder.withInputBatchCount(DatadogEventWriter.java:463)
> 2023-07-18T22:09:53.1693200Z 	at com.google.cloud.teleport.datadog.DatadogIO$Write.expand(DatadogIO.java:86)
> 2023-07-18T22:09:53.1693719Z 	at com.google.cloud.teleport.datadog.DatadogIO$Write.expand(DatadogIO.java:60)
> 2023-07-18T22:09:53.1694216Z 	at org.apache.beam.sdk.Pipeline.applyInternal(Pipeline.java:545)
> 2023-07-18T22:09:53.1695084Z 	at org.apache.beam.sdk.Pipeline.applyTransform(Pipeline.java:496)
> 2023-07-18T22:09:53.1695584Z 	at org.apache.beam.sdk.values.PCollection.apply(PCollection.java:365)
> 2023-07-18T22:09:53.1696152Z 	at com.google.cloud.teleport.templates.PubSubToDatadog.run(PubSubToDatadog.java:222)
> 2023-07-18T22:09:53.1696847Z 	at com.google.cloud.teleport.templates.PubSubToDatadog.main(PubSubToDatadog.java:150)
> 2023-07-18T22:09:53.1697010Z 	... 5 more
> ```
> 
> In the smoke tests, going to rebase/rerun to double check

Thanks! I just pushed up a change to make `batchCount=10` since that's the minimum and can't be changed from template params. Trying to get tests running locally to make sure that works. 🙏

ash-ddog commented:
@bvolpato Just checking in. This has been sitting a while. Anything required on our end to get this in? 🙏

bvolpato commented:
Hi Ash! Sorry for lack of feedback here. What's going on is that we are moving the `it` module, which is a larger chunk of the PR, to Beam (https://github.com/apache/beam/tree/master/it).

@pranavbhandari24 has been working in this transition.

For context, Datadog has an IO only in the DataflowTemplates, and resource managers that follow our general interface, but are faking/mocking a server, due to Datadog not having any containers.

Do we expect to have RMs sitting in the repository or get rid of all `it`? At least transitively, should we push here and move to Beam at a later time?

bvolpato commented:
Hey Ash,

Do you mind re-organizing imports? It's mostly `com.google.cloud.teleport.it` -> `org.apache.beam.it`

If you want I can take care of it in your fork too, just give me write permission!

ash-ddog commented:
@bvolpato No worries at all! Just wanted to make sure the ball was not on my end.

I've added you as a contributor on my custom fork, just in case you need to do work in the future.

But I've rebased this PR on `main` and fixed imports. Will check out any test failures. Thanks for helping move this forward! 👍

bvolpato reviewed (approved):
Really good stuff, thanks! 

The last round of tests was affected by some sort of unavailability on GitHub

> Error: Response status code does not indicate success: 502 (Bad Gateway).

I will wait for it to finish and proceed with imports.

bvolpato commented:
@ash-ddog Sorry for being slow here. There's one thing, need to mock `.withLogConsumer` in the TestContainers implementation, something like:

```
doReturn(container).when(container).withLogConsumer(any());
```

I would have done it, but my invitation expired :-( Can you add me again? I will try to do it asap.

ash-ddog commented:
@bvolpato No worries, it happens! My bad for not seeing this! Just added you as a collaborator again, let me know if I can help in getting this in! 🙏
